File: src/main/java/Problem0684.java
import java.math.BigInteger;

public class Problem0684 {

    private final int[] oneToNine = {0, 1, 3, 6, 10, 15, 21, 28, 36, 45};

    public Problem0684() {
        long startTime = System.nanoTime();
        long result = 0;

        // Solution:
        long[] fibo = this.findFibos(90);
        float i = 1;
        for (long f : fibo) {
            result += this.bigS(f) % 1_000_000_007;
            System.out.println(i++ / 90 * 100 + " %");
//            i++;
        }
        //TODO finish

        long timeToResolve = System.nanoTime() - startTime;
        System.out.println("Result:\t" + result + "\tTime:\t" + (((double) timeToResolve / 1_000_000) > 1000 ?
                (((double) timeToResolve / 1_000_000_000) + "s") :
                (((double) timeToResolve / 1_000_000) + "ms")));
    }

    private long[] findFibos(int n) {
        long[] result = new long[n];

        result[0] = 1;
        result[1] = 1;
        for (int i = 2; i < n; i++) {
            result[i] = result[i - 1] + result[i - 2];
        }

        return result;
    }

    private BigInteger findSmallestDigitSum(long n) {
        return new BigInteger(n % 9 + "9".repeat((int) n / 9));
    }

    private long bigS(long n) {
        BigInteger result = new BigInteger("0");

        String temp = n + "";

        int firstNumberOfNines = 0;
        if (temp.length() > 1) {
            firstNumberOfNines = Integer.parseInt("1".repeat((n + "").length() - 2) + "0");
        }

        int nextNumberOfNines = 0;
        if (temp.length() > 2) {
            for (int i = 1; i <= temp.length() - 2; i++) {
                nextNumberOfNines += Integer.parseInt("9".repeat(i));
            }
        }

        while (n > 0) {
            String test = n + "";
            if (test.length() == 1) {
                result = result.add(BigInteger.valueOf(this.oneToNine[(int) n]));
                n -= n;
            } else {
                result = result.add(new BigInteger("5" + "9".repeat(firstNumberOfNines) + "3" +
                        ("9".repeat(nextNumberOfNines)).repeat(Integer.parseInt(test.substring(0, 1))) +
                        ("9".repeat(test.length() - 1)).repeat(Integer.parseInt(test.substring(0, 1)) - 1) +
                        "0".repeat(test.length() - 2) + "1"));
                if (firstNumberOfNines > 0)
                    firstNumberOfNines = Integer.parseInt((firstNumberOfNines + "").substring(1));
                if (test.length() > 2)
                    nextNumberOfNines -= Integer.parseInt("9".repeat(test.length() - 2));

                n -= (test.charAt(0) - 48) * Integer.parseInt("1" + "0".repeat(test.length() - 1));
            }
        }
        return Long.parseLong(result.mod(new BigInteger("1000000007")).toString());
    }

}

File: src/main/cs/Problem0684.cs
using System;
using System.Diagnostics;
using System.Numerics;
using System.Reflection.Emit;
using DocumentFormat.OpenXml.Spreadsheet;
using Euler.test.cs;
using Microsoft.VisualBasic;

namespace Euler.main.cs
{
    public class Problem0684
    {
        private long mod = 1_000_000_007;

        public Problem0684()
        {
            Stopwatch stopWatch = new Stopwatch();
            stopWatch.Start();
            long result = 0;

            // Solution:
            ulong[] fibo = Fibo(90);
            for (int k = 2; k <= 90; k++)
            {
                result += ModS(fibo[k]);
                result %= mod;
            }

            stopWatch.Stop();
            string elapsedTime = stopWatch.Elapsed.ToString();
            Console.WriteLine("Result:\t" + result + "\tTime:\t" +
                              (double.Parse(elapsedTime.Substring(elapsedTime.LastIndexOf(":") + 1, 2)) >= 1
                                  ? double.Parse(elapsedTime.Substring(elapsedTime.LastIndexOf(":") + 1)) + " s"
                                  : double.Parse(elapsedTime.Substring(elapsedTime.IndexOf(".") + 1)) / 10_000 +
                                    " ms"));
            if (Test.DoBenchmark)
            {
                Benchmark.AddTime(684, double.Parse(elapsedTime.Substring(elapsedTime.IndexOf(".") + 1)) / 10_000);
            }
        }

        private long ModS(ulong k)
        {
            int n = (int) (k / 9);
            BigInteger sum = Mod(2 * (BigInteger.ModPow(2, n, mod) * BigInteger.ModPow(5, n + 2, mod) - 7) - 9 * n);

            for (ulong r = k % 9 + 2; r <= 9; r++)
            {
                sum -= Mod(r * BigInteger.ModPow(10, n, mod) - 1);
            }

            return (long) Mod(sum);
        }

        private ulong[] Fibo(ulong n)
        {
            n++;
            ulong[] result = new ulong[n];
            result[0] = 0;
            ulong a = result[1] = 1;
            ulong b = result[2] = 1;
            for (ulong i = 3; i < n; i++)
            {
                ulong help = b;
                b += a;
                a = help;
                result[i] = b % (ulong) mod;
            }

            return result;
        }

        private BigInteger Mod(BigInteger n)
        {
            BigInteger nMod = n % mod;
            return nMod >= 0 ? nMod : nMod + mod;
        }
    }
}